### ISSUE TITLE
Proposal for update

### ISSUE THREAD

**Issue by pa36651** [M0r13n/pyais#88]
I use this code to decode nmea0183 data (!AIVD...) collected from the AIS-reciever in my boat. Sometimes the message get errors. When the pyais try to decode the corrupted AIS-message, the code stop with a message as follows.
Traceback (most recent call last):

  File "C:\Users\Asus\OneDrive\Documents\__Python\decode_AIS\decode_AIS.py", line 93, in <module>
    for msg in FileReaderStream(filename):

  File "C:\Users\Asus\anaconda3\lib\site-packages\pyais\stream.py", line 54, in _assemble_messages
    msg: NMEAMessage = NMEAMessage(line)

  File "C:\Users\Asus\anaconda3\lib\site-packages\pyais\messages.py", line 218, in __init__
    self.bit_array: bitarray = decode_into_bit_array(self.payload, self.fill_bits)

  File "C:\Users\Asus\anaconda3\lib\site-packages\pyais\util.py", line 34, in decode_into_bit_array
    raise ValueError(f"Invalid character: {chr(c)}")

ValueError: Invalid character: !

The above error come with the message: !AIVDM,1,1,,B,3815;`100!PhmnPPwL=3OmUd0Dg:,0*45

I would prefer if errors like this one doesn't stop the execution, and instead just give a notice that there was a message with an error, and the continue the process of the next AIS message.

**Comment by M0r13n** [comment]
Hey @pa36651 ,

technically, your message is invalid, because the payload contains a `!` character. This is not allowed according to the NMEA 0183 specification. Therefore, this causes an error. 

But I think that it is safe to ignore this error. True to the motto "Be liberal in what you accept", I think that we should not raise an error in these situations. Only non printable characters should result in an exception. In order to make the exception more useful, I decided to add a special `NonPrintableCharacterException`. This is way more specific than a simple `ValueError`.

Therefore, your exception won't occur anymore once you update to version **2.2.3** of pyais.

<hr>

Regarding your actual question:

> I would prefer if errors like this one doesn't stop the execution, and instead just give a notice that there was a message with an error, and the continue the process of the next AIS message.

This is up to you as a user of the library to decide. Just wrap the necessary method in a try/except block. If you are reading messages from a file, the code for that could look like this:

```py
import pathlib

from pyais.stream import FileReaderStream
from pyais.exceptions import NonPrintableCharacterException

filename = pathlib.Path(__file__).parent.joinpath('sample.ais')

for msg in FileReaderStream(str(filename)):
    try:
        decoded = msg.decode()
    except NonPrintableCharacterException:
        print('could not decode:', msg)
```

**Comment by pa36651** [comment]
Thank you! 
I find the pyais very nice!

**Comment by M0r13n** [comment]
Nice, thank you! 👍

**Comment by pa36651** [comment]
I updated to 2.2.3 of pyais. I'm sorry to say that the issue seems to remain. I added the proposed code with try/except, but the error is set already at the line `for msg in FileReaderStream(str(filename)):`

Attached some logg-files that I want to decode (SXKH22_3.tx, SXKH22_4.txt), and my code (decode_AIS_test.py).
The thing is that at least my AIS transponder doesn't seem to do any filtering om the recieved messages, It just pass whatever comes in, as nmea0183 out. The logg-file is the complete nmea0183 bus logg representing approx 6 hours of sailing.
[_SXKH22_3.txt](https://github.com/M0r13n/pyais/files/9946587/_SXKH22_3.txt)
[_SXKH22_4.txt](https://github.com/M0r13n/pyais/files/9946551/_SXKH22_4.txt)
[decode_AIS_test.txt](https://github.com/M0r13n/pyais/files/9946553/decode_AIS_test.txt)

The error message:
Traceback (most recent call last):

  File "C:\Users\Asus\OneDrive\Documents\__Python\decode_AIS\decode_AIS_test.py", line 75, in <module>
    for msg in FileReaderStream(str(filename)):

  File "C:\Users\Asus\anaconda3\lib\site-packages\pyais\stream.py", line 54, in _assemble_messages
    msg: NMEAMessage = NMEAMessage(line)

  File "C:\Users\Asus\anaconda3\lib\site-packages\pyais\messages.py", line 218, in __init__
    self.bit_array: bitarray = decode_into_bit_array(self.payload, self.fill_bits)

  File "C:\Users\Asus\anaconda3\lib\site-packages\pyais\util.py", line 35, in decode_into_bit_array
    raise NonPrintableCharacterException(f"Non printable character: '{hex(c)}'")

NonPrintableCharacterException: Non printable character: '0x10'

**Comment by M0r13n** [comment]
> I'm sorry to say that the issue seems to remain

Not quite. If you look closely, you can see that the error slightly differs.

Nevertheless, this is undesired behavior.  I fixed it with v2.2.4. :-)
